libraries/Microsoft.Bot.Builder.Dialogs/Dialog.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    /// <summary>
+    /// Abstract class for all Dialog objects that can be added to a `DialogSet`. The dialog should generally
+    /// be a singleton and added to a dialog set using `DialogSet.add()` at which point it will be 
+    /// assigned a unique ID.
+    /// </summary>
+    public abstract class Dialog
+    {
+        /// <summary>
+        /// Method called when a new dialog has been pushed onto the stack and is being activated.
+        /// </summary>
+        /// <param name="dc">The dialog context for the current turn of conversation.</param>
+        /// <param name="dialogArgs">(Optional) arguments that were passed to the dialog during `begin()` call that started the instance.</param>  
+        public abstract Task DialogBegin(DialogContext dc, object dialogArgs = null);
+
+        /// <summary>
+        /// Indicates whether the class supports DialogContinue
+        /// </summary>
+        public abstract bool HasDialogContinue { get; }
+
+        /// <summary>
+        /// (Optional) method called when an instance of the dialog is the "current" dialog and the 
+        /// user replies with a new activity. The dialog will generally continue to receive the users 
+        /// replies until it calls either `DialogSet.end()` or `DialogSet.begin()`.
+        /// If this method is NOT implemented then the dialog will automatically be ended when the user replies.
+        /// </summary>
+        /// <param name="dc">The dialog context for the current turn of conversation.</param>
+        public virtual Task DialogContinue(DialogContext dc)
+        {
+            throw new NotImplementedException();
```

Comment:
This doesn't make any sense. 

There are 2 abstract methods, and 2 that throw. The comment says "if not implemented, the dialog will be ended". That's NOT what this code does. This code throw an exception. 

This design has to change.

Reply:
He's trying to model the fact that Dialog has optional methods in my implementation which is **By Design**.  There is behavior (which I'm not willing to remove) that's keyed off whether you've implemented a given method or not.  If you want to get into the details of how the dialog stack works we can at some point but the current design is based off 2+ years of me evolving the waterfall and dialog stack pattern. So for now just accept that it HAS to work the way it works.